logstash-core/src/main/java/org/logstash/Event.java

```diff
@@ -213,6 +240,62 @@ public void setField(final FieldReference field, final Object value) {
         }
     }
 
+    private boolean isTagsWithMap(final FieldReference field) {
+        return field.getPath() != null && field.getPath().length > 0 && field.getPath()[0].equals(TAGS);
+    }
+
+    private boolean isLegalTagValue(final Object value) {
+        if (value instanceof String || value instanceof RubyString || value == null) {
+            return true;
+        } else if (value instanceof List) {
+            for (Object item: (List) value) {
+                if (!(item instanceof String) && !(item instanceof RubyString)) {
+                    return false;
+                }
+            }
+            return true;
+        }
+
+        return false;
+    }
+
+
+    private void setTagsAndFailTags(final Object value) {
+        tag(TAGS_FAILURE_TAG);
+        tagFailTags(Valuefier.convert(value));
+    }
+
+    /**
+     * handle key/val pair in `tags`, add _tagsparsefailure to `tags`.
+     * rename FieldReference from `tags` to `_tags`. If existing `_tags` is a list, append value to `_tags`,
+     * eg. [tags][key] -> [_tags][list_index][key]
+     * @param field FieldReference of `tags` with map path
+     * @param value
+     */
+    private void setTagsAndFailTagsWithMap(final FieldReference field, final Object value) {
+        tag(TAGS_FAILURE_TAG);
+
+        List<String> paths = new ArrayList<>();
+        paths.add(TAGS_FAILURE);
+
+        // take list size of _tags as index
+        final Object failTags = Accessors.get(data, TAGS_FAILURE_FIELD);
+        int index = (failTags instanceof List)? ((List) failTags).size() : 0;
+        paths.add(Integer.toString(index));
+
+        // rebuild path & key
+        for(int i = 1; i < field.getPath().length; i++) {
+            paths.add(field.getPath()[i]);
+        }
```

Comment:
`subList` from 1 is correct as this method handles Map which must contain at least one path, otherwise `subList(1, path)` throw index out of bound exception. `for 1..length` guarantees eligible index. I prefer to keep it.